### ISSUE TITLE
No way to specify flex-basis css property in tailwind

### ISSUE THREAD

**Issue by connecteev** [tailwindcss/tailwindcss#603]
I am trying to create a responsive flexbox grid, like https://codepen.io/astrotim/pen/WQwqbW
which requires me to specify 2 styles for flex-basis (both of which are important, one does not override the other).
```
.flex-grid-item {
	height: auto;
	flex-grow: 1;
	flex-shrink: 0;
	flex-basis: 33.3333%;
	flex-basis: 200px;
	display: flex;
}
```

Looking at the generated tailwind.css, there doesnt seem to be a way to specify these styles in the tailwind format, unless you define your own css class.

**Comment by benface** [comment]
> both of which are important, one does not override the other

Not sure why you think that; the second `flex-basis` declaration does indeed override the first one. Try removing the first one and you should see no difference whatsoever in the result.

**Comment by adamwathan** [comment]
In this case you can just use `width` as well, which does exactly what `flex-basis` does in this example:

https://codepen.io/anon/pen/BvwEYZ

We didn't provide flex-basis utilities originally because width does the same thing for 99.9% of cases and we didn't want to increase the size of the framework, but you could easily add your own flex-basis utilities by following the documentation:

https://tailwindcss.com/docs/adding-new-utilities

You could write a plugin for it too:

https://tailwindcss.com/docs/plugins#adding-utilities

**Comment by connecteev** [comment]
@adamwathan thanks for the response.

**Comment by homerjam** [comment]
I'd vote for at least a very limited implementation to be included so that it can easily be extended in the theme settings. I seem to hit the 0.1% of cases rather frequently!

**Comment by adamwathan** [comment]
@homerjam What's an example of a situation where you need it? I've still not needed it even once in three years of building stuff with Tailwind.

**Comment by homerjam** [comment]
@adamwathan 😳  I figured out where I've been going wrong while I was trying to create a demo. I'd like to retract my previous comment! 😄

**Comment by HamzaAmar** [comment]
@adamwathan  Hello adam some use case when you like the flex item to be in the some size for both direction column and row without add an extra media query

**Comment by mattbloomfield** [comment]
I just ran into a use case for this - I was building a tabbed-view component and wanted to force the tabs to be equal width. 

Because it's a component I don't know how many tabs it will need, so I can't know the width. 

`class="flex-grow" and style="flex-basis: 0"` on each child was a perfect solution. 

What would you recommend here, @adamwathan ? 

Try it out here: https://play.tailwindcss.com/T21pBg0Sum

**Comment by marie-maxime** [comment]
@mattbloomfield  you can use`flex-1`, which will output `flex: 1 1 0%;`. See https://tailwindcss.com/docs/flex

**Comment by Yihao-G** [comment]
There is a big difference, and I just encountered a similar case like this. 

See the example https://play.tailwindcss.com/ikHCIqWZcw

Adjust the width of the viewport and see how items are wrapped. Using `width` will never wrap, but `flex-basis` will if the content can no longer fit.

**Comment by adamwathan** [comment]
Oh super interesting, first time in 5 years of flexbox I have seen a demo that shows a difference! I can't find anything in the spec or any tutorials that can explain this difference in behavior, seen anything that explains it?

**Comment by Yihao-G** [comment]
I guess that related to the different behaviours between `width: 100%` and `width: auto`. Just to provide a background to explain why, see the characteristics described in [this video](https://www.youtube.com/watch?v=-st14lUQD3U): 

> `width: auto` means use whatever space is available without overflowing

-----------------------

So in the example above, when the `width` is set to a fixed value (`33.3333%`), the width of the element is calculated based on the parent element width and will try to not change it. Thus, the content within the element overflows. When we remove the fixed width, the `width` property becomes the default value `auto` and the browser will try to not overflow the element.

`flex-basis`, however, in my understanding, is telling the browser the **suggested** width - weaker than the `width` property. In the case of the example, the browser will follow the `flex-basis` value as the element width, only if it can do so<sup>*</sup>. In the second case of the example, The element `width` is `auto`, which means, it tries to not overflow. When the viewport width reduces, the element no longer fits the text, and then `flex-wrap` kicks in, to make the content within the element not overflowing.

~~Hope this doesn't make things more confusing...~~

-----------------------

<sup>*</sup>: Not exactly the definition of `flex-basis`, but you can try to think `flex-basis` in this way, along with the "official" description of "[the initial main size of a flex item](https://developer.mozilla.org/en-US/docs/Web/CSS/flex-basis)".

**Comment by benface** [comment]
Note that if you set `min-width` to `0` (the default value is `auto`) in the second example, it will behave like the first one (it won't wrap and the content will overflow).

**Comment by geoyws** [comment]
I've been using `flex-basis` for many years now and only started using Tailwind recently. It'd be nice if others like me could hit the ground running using what we're used to 😄  and not have to come here to find out why it's not listed in the docs 😅 . Gonna continue to love using Tailwind without it regardless 😃 👍🏼 .

**Comment by ianhernandez** [comment]
Hey Adam @adamwathan I know there's a bunch going on considering the latest CLI/JIT updates. 
I'm wondering if we can we open this up again with the intent of bringing native `flex-basis` support via JIT compilation.

I haven't had any reason to use another styling library, but I've come to a point where I have to decide how to go about implementing Emotion into our design system just to be able to handle this particular property with more flexibility than inline styling.

**Comment by adamwathan** [comment]
Definitely not opposed to adding it at this point — what's a specific situation where you aren't able to use `width` though? They are _very_ rare so you might be fine with width until we get this in for the next minor release (which is probably a couple of months out).

**Comment by nayaabkhan** [comment]
I find the `flex-basis` useful when I want the item to use the same dimension irrespective of the direction of the flex. E.g., in a card with a photo on the left (basis = width) or top (basis = height) depending on the available space. The axis of dimension is implicit in such cases.

**Comment by jhardbower** [comment]
`width` can work most of the time, but becomes an issue on small screens where the set width of the element is larger than the device.

Design implemented with flex-basis: https://play.tailwindcss.com/JEgnodUwXH?file=config
Design implemented with width: https://play.tailwindcss.com/V2SBQ0cceR?file=config

Shrink the width of the viewport down and you'll see it overflows with `width`.

**Comment by alexfinnarn** [comment]
I came here wondering why Tailwind had no `flex-basis` utility just like most other people before me. I know adding `flex-basis` fits into the 80/20 use case rule and can understand why including it seemed like overkill initially. 

TLDR; While using `width` will work 99.9% of the time, it is technically incorrect and `flex-basis` should be used in all cases according to some CSS gurus. 

Based on my knowledge, on principle, you should not be using width within Flexbox. Width is a finite measurement whereas Flexbox is anything but finite. The whole point is to "flex" content within a container around other content and allow the browser to do the hard work of lining things up visually. 

If you say `width: 25%` for an element, then you get the edge cases where a width of 25% doesn't work with certain page widths and other elements surrounding it. If you say `flex-basis: 25%` instead, the browser can line up items so that the element is actually 27% of the container in some cases but visually, it aligns with the rest of the content and looks better for users. I think most people don't encounter this problem since they make sure their breakpoints look good without analyzing all screen sizes and possible element arrangments...which I don't do nor should most people spend their time doing.

None of this perspective comes from me, though. I credit Frontend Masters and Jen Kramer when she said this in a course summing up how to define widths within Flexbox.

> Never use the width property on your flex items. Do not use the width property on your flex items. You're gonna use flex-basis, okay? Flex-basis.

That logic makes sense to me..and kind of resembles a mob boss coercing someone, lol...so I'd prefer to always use `flex-basis` instead of `width` while using Flexbox.

I hope this explanation helps in some way, but I wanted to add it since someone who I think is a CSS guru categorically opposes the other mentioned solutions on principle even if they work.
